src/table.ts

```diff
@@ -778,133 +749,46 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
-              startValue as string,
-              lastRowKey as string
-            );
-          const endKeyIsNotRead =
-            !endValue ||
-            (endValue as Buffer).length === 0 ||
-            TableUtils.lessThan(lastRowKey as string, endValue as string);
-          if (startKeyIsRead) {
-            if (endKeyIsNotRead) {
-              // EndKey is not read, reset the range to start from lastRowKey open
-              range.start = {
-                value: lastRowKey,
-                inclusive: false,
-              };
-            } else {
-              // EndKey is read, remove this range
-              ranges.splice(index, 1);
-            }
-          }
-        }
-
-        // Remove rowKeys already read.
-        rowKeys = rowKeys.filter(rowKey =>
-          TableUtils.greaterThan(rowKey, lastRowKey as string)
-        );
-
-        // If there was a row limit in the original request and
-        // we've already read all the rows, end the stream and
-        // do not retry.
-        if (hasLimit && rowsLimit === rowsRead) {
-          userStream.end();
-          return;
-        }
-        // If all the row keys and ranges are read, end the stream
-        // and do not retry.
-        if (rowKeys.length === 0 && ranges.length === 0) {
-          userStream.end();
-          return;
+      userStream.end = (chunk?: any, encoding?: any, cb?: () => void) => {
+        rowStreamUnpipe(rowStream, userStream);
+        userCanceled = true;
+        if (activeRequestStream) {
+          activeRequestStream.abort();
         }
-      }
-
-      // Create the new reqOpts
-      reqOpts.rows = {};
-
-      // TODO: preprocess all the keys and ranges to Bytes
-      reqOpts.rows.rowKeys = rowKeys.map(
-        Mutation.convertToBytes
-      ) as {} as Uint8Array[];
-
-      reqOpts.rows.rowRanges = ranges.map(range =>
-        Filter.createRange(
-          range.start as BoundData,
-          range.end as BoundData,
-          'Key'
+        return originalEnd(chunk, encoding, cb);
+      };
+      const chunkTransformer: ChunkTransformer = new ChunkTransformer({
+        decode: options.decode,
+      } as any);
+
+      const strategy = new ReadRowsResumptionStrategy(
+        chunkTransformer,
+        options,
+        Object.assign(
+          {tableName: this.name},
+          this.bigtable.appProfileId
+            ? {appProfileId: this.bigtable.appProfileId}
+            : {}
         )
       );
 
-      if (filter) {
-        reqOpts.filter = filter;
-      }
+      // TODO: Consider removing populateAttemptHeader.
+      const gaxOpts = populateAttemptHeader(0, options.gaxOptions);
 
-      if (hasLimit) {
-        reqOpts.rowsLimit = rowsLimit - rowsRead;
+      // Attach retry options to gax if they are not provided in the function call.
+      gaxOpts.retry = strategy.toRetryOptions(gaxOpts);
+      if (gaxOpts.maxRetries === undefined) {
+        gaxOpts.maxRetries = maxRetries;
       }
```

Comment:
This is done to maintain existing behaviour regarding setting maxRetries using the existing line in table.ts that says `const maxRetries = is.number(this.maxRetries) ? this.maxRetries! : 10;`. However, if the user supplies maxRetries in gaxOptions to the call to createreadstream then the maxRetries value that the user supplied for the call will be used.